From: Andreas Schuh <andreas.schuh.84@gmail.com>
Date: Thu, 1 May 2014 20:16:16 +0100
Subject: [PATCH] Set VERSION property of library targets to <major>.<minor>.


diff --git a/CMakeLists.txt b/CMakeLists.txt
index 54b5c35..756ddba 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -34,7 +34,7 @@ version_numbers (
     PACKAGE_VERSION_PATCH
 )
 
-set (PACKAGE_SOVERSION "${PACKAGE_VERSION_MAJOR}")
+set (PACKAGE_SOVERSION "${PACKAGE_VERSION_MAJOR}.${PACKAGE_VERSION_MINOR}")
 
 # ----------------------------------------------------------------------------
 # options
